pytorch_lightning/callbacks/model_checkpoint.py

```diff
@@ -239,6 +243,22 @@ def on_validation_end(self, trainer, pl_module) -> None:
             return
         self.save_checkpoint(trainer)
 
+    def on_train_end(self, trainer, *args, **kwargs) -> None:
+        """
+        checkpoints can be saved at the end of the trianing
+        """
+        if not self._trigger_on_train_end:
+            return
+        # as we advance one step at end of training, we use global_step - 1
+        # to avoid saving duplicates
+        trainer.global_step -= 1
+        if (not self._should_skip_saving_checkpoint(trainer) and trainer.checkpoint_connector.has_trained):
+            if self.save_last and self.verbose:
+                rank_zero_info("Saving last checkpoint...")
+            monitor_candidates = self._monitor_candidates(trainer)
+            self._save_last_checkpoint(trainer, monitor_candidates)
+        trainer.global_step += 1
+
```

Comment:
```suggestion
    def on_train_end(self, trainer, pl_module) -> None:
        """Save a checkpoint at the very end of training.

        This will only save a checkpoint if `save_last` is also enabled
        as the monitor metrics produced by training or validation steps or end of epochs
        is not guaranteed to be available at this stage.
        """
        if self._should_skip_saving_checkpoint(trainer) or not trainer.checkpoint_connector.has_trained:
            return

        initial_save_last = self.save_last
        if self._save_on_train_end and not self.save_last:
            rank_zero_warn(
                "Requested to save a checkpoint at the end of training but save_last is not set. Temporarily setting save_last=True to save."
            )
            self.save_last = True
        if self.verbose:
            rank_zero_info("Saving last checkpoint...")

        # as we advance one step at end of training, we use global_step - 1
        # to avoid saving duplicates
        trainer.global_step -= 1
        monitor_candidates = self._monitor_candidates(trainer)
        self._save_last_checkpoint(trainer, monitor_candidates)
        trainer.global_step += 1
        self.save_last = initial_save_last

```

what do you think of this? 

also what should happen if save_last is not set to True? should save on train end take precedence and temporarily override it? should we move the save_last check out of _save_last_checkpoint so the property needs to be checked first before we call save_last_checkpoint?

Reply:
i think the original thought is `save_on_train_end` is dependent on `save_last`, so only enabled when `save_last` is set also. What you proposed is to always enable is regardless of `save_last`. To make `save_on_train_end` as an independent triggering, makes sense also.

Reply:
@carmocca @awaelchli what do you think?

Reply:
I think I prefer the current implementation, maybe throwing a warning so people know they should set both.